### PR TITLE
feat(repo/stats): Log renovate repo statistics

### LINKED ISSUE #15309: Improve overall renovate repo statistics logging
### What would you like Renovate to be able to do?

Log overall statistics per repository, i.e. total of update PR offered, merged, closed etc.

### If you have any ideas on how this should be implemented, please tell us here.

Use current cache to retrieve and then process current PR list for a given repo.  

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
Log renovate overall statistics for the current repo run.

## Context
Closes #15309

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by Gabriel-Ladzaretti** [comment]
Below is the tail of the loggings for https://github.com/ladzaretti/config-onboarding/pulls.


```
DEBUG: getPrList success (repository=ladzaretti/config-onboarding)
       "pullsTotal": 4,
       "requestsTotal": 1,
       "apiQuotaAffected": true
 INFO: Renovate repository PR statistics (repository=ladzaretti/config-onboarding)
       "repo": "ladzaretti/config-onboarding",
       "stats": {"created": 3, "open": 1, "closed": 1, "merged": 1}
DEBUG: Repository timing splits (milliseconds) (repository=ladzaretti/config-onboarding)
       "splits": {"init": 6329, "extract": 1366, "lookup": 144, "update": 8125},
       "total": 16595
DEBUG: http statistics (repository=ladzaretti/config-onboarding)
       "urls": {
         "https://api.github.com/graphql (POST,200)": 2,
         "https://api.github.com/repos/ladzaretti/config-onboarding/branches/main/protection (GET,404)": 1,
         "https://api.github.com/repos/ladzaretti/config-onboarding/issues/4/comments (GET,200)": 1,
         "https://api.github.com/repos/ladzaretti/config-onboarding/issues/4/comments (POST,201)": 1,
         "https://api.github.com/repos/ladzaretti/config-onboarding/pulls (GET,200)": 2
       },
       "hostStats": {"api.github.com": {"requestCount": 7, "requestAvgMs": 545, "queueAvgMs": 0}},
       "totalRequests": 7
 INFO: Repository finished (repository=ladzaretti/config-onboarding)
       "durationMs": 16595
DEBUG: Renovate exiting
Done in 33.61s.

**Comment by rarkins** [comment]
Debug is preferred instead of into. No need for repo field as we already log repository. Is the created field just the total of the other ones?

**Comment by Gabriel-Ladzaretti** [comment]
>  Is the created field just the total of the other ones?
yes, excluding onboarding

**Comment by Gabriel-Ladzaretti** [comment]
updated as per https://github.com/renovatebot/renovate/pull/15308#issuecomment-1109636732

```
DEBUG: Renovate repository PR statistics (repository=ladzaretti/config-onboarding)
       "stats": {"created": 3, "open": 1, "closed": 1, "merged": 1}

```

File: lib/workers/repository/finalise/repository-statistics.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import { logger } from '../../../logger';
+import { getPrCache } from '../../../modules/platform/github/pr';
+import { PrState } from '../../../types';
+import { GithubHttp } from '../../../util/http/github';
+
+const githubApi = new GithubHttp();
+const renovateBot = 'renovate[bot]';
+
+type PrStats = {
+  created: number;
```
Comment (nabeelsaabna) [inline lib/workers/repository/finalise/repository-statistics.ts:11]: update to `total`

**Review by nabeelsaabna — DISMISSED** [review]

File: lib/workers/repository/finalise/repository-statistics.spec.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  mockedFunction,
+} from '../../../../test/util';
+import { logger } from '../../../logger';
+import { getPrCache } from '../../../modules/platform/github/pr';
+import { runRenovateRepoStats } from './repository-statistics';
+
+jest.mock('../../../modules/platform/github/pr');
+jest.mock('../../../util/http/github');
+
+const prCache = loadJsonFixture('./pr-cache.json');
```
Comment (viceice) [inline lib/workers/repository/finalise/repository-statistics.spec.ts:14]: please use fixtures class

File: lib/workers/repository/finalise/repository-statistics.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      case PrState.Merged:
+        prStats.merged += 1;
+        break;
+      case PrState.Closed:
+        prStats.closed += 1;
+        break;
+      case PrState.Open:
+        prStats.open += 1;
+        break;
+      // istanbul ignore next: exclude PrState.All & .NotOpen
```
Comment (viceice) [inline lib/workers/repository/finalise/repository-statistics.ts:54]: why not simply add appropriate PR to samples?

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/repository/finalise/repository-statistics.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    closed: 0,
+    merged: 0,
+  };
+}
+
+export async function runRenovateRepoStats(
+  config: RenovateConfig
+): Promise<void> {
+  const prStats = initStats();
+  const prCache = await getPrCache(
```
Comment (zharinov) [inline lib/workers/repository/finalise/repository-statistics.ts:30]: I'm not sure if it's okay to rely on it.
- It's platform-dependent and GitHub-only for now
- Its structure is subject to change
- Users may want to see *delta* values, i.e. how many PRs are open/closed/merged for this particular run. Seems to be complicated to implement with repository cache.

Most likely, it won't be much harder to init/update stats counters manually after relevant platform actions. Something similar is already done for HTTP statistics.

File: lib/workers/repository/finalise/repository-statistics.spec.ts (hunk 2 above)
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/finalise/repository-statistics.spec.ts:14]: this is an renovate specific util function, are they all redundant? as i was also requested to not use loadFixture but rather use Fixture class instead

File: lib/workers/repository/finalise/repository-statistics.ts (hunk 4 above)
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/finalise/repository-statistics.ts]: it is intended for github use only, ive added the required checks before calling the newly added function. thanks for the heads up.

you might be right, will wait for some further instruction.

other than that, ive added some more changes to files related to github which are needed for this to work. can u please check out https://github.com/renovatebot/renovate/pull/15319

Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/finalise/repository-statistics.ts]: as @MaronHatoum found and pointed out, on the first run of renovate the pr are filtered and storted by config.renovateusername then stored with repo as key.

on the second run we are getting 304 code and the previous cahced list is returned.
therefore we need to add support for pr created by [bot] and retrieve them later on

**Pull request closed without merging**